components/cdc/src/observer.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::{Arc, RwLock};
+
+use raft::StateRole;
+use tikv::raftstore::coprocessor::*;
+use tikv::raftstore::Error as RaftStoreError;
+use tikv_util::collections::HashSet;
+use tikv_util::worker::Scheduler;
+
+use crate::endpoint::Task;
+use crate::Error as CdcError;
+
+/// An Observer for CDC.
+///
+/// It observes raftstore internal events, such as:
+///   1. Raft role change events,
+///   2. TODO Apply command events.
+#[derive(Clone)]
+pub struct CdcObserver {
+    sched: Scheduler<Task>,
+    // A shared registry for managing observed regions.
+    observe_regions: Arc<RwLock<HashSet<u64>>>,
```

Comment:
Everytime a region changes its role it will be locked. Is that acceptable?

Reply:
I am afraid this is the only way to notify the CDC in time, but it's definitely worth to improve later.